File: atividade2/merche.java
package atividade2;

import java.io.*;
import java.nio.file.*;
import java.util.ArrayList;
import java.util.List;

public class merche {

    // Método para realizar o Merge Sort
    public static void mergeSort(List<Integer> arr) {
        if (arr.size() < 2) {
            return; // Lista já está ordenada
        }

        int mid = arr.size() / 2;

        // Dividir a lista
        List<Integer> left = new ArrayList<>(arr.subList(0, mid));
        List<Integer> right = new ArrayList<>(arr.subList(mid, arr.size()));

        // Ordenar cada metade
        mergeSort(left);
        mergeSort(right);

        // Mesclar as metades ordenadas
        merge(arr, left, right);
    }

    private static void merge(List<Integer> arr, List<Integer> left, List<Integer> right) {
        int i = 0, j = 0, k = 0;

        // Mesclando os dois arrays
        while (i < left.size() && j < right.size()) {
            if (left.get(i) <= right.get(j)) {
                arr.set(k++, left.get(i++));
            } else {
                arr.set(k++, right.get(j++));
            }
        }

        // Adicionando os elementos restantes
        while (i < left.size()) {
            arr.set(k++, left.get(i++));
        }
        while (j < right.size()) {
            arr.set(k++, right.get(j++));
        }
    }

    // Método principal para medir tempo e executar a lógica
    public static void measureMemoryAndTime(String filePath) throws IOException {
        // Ler números do arquivo
        List<Integer> numbers = new ArrayList<>();
        for (String line : Files.readAllLines(Paths.get(filePath))) {
            numbers.add(Integer.parseInt(line.trim()));
        }

        // Medir tempo inicial
        long startTime = System.nanoTime();

        // Realizar o Merge Sort
        mergeSort(numbers);

        // Medir tempo final
        long endTime = System.nanoTime();
        double executionTimeMs = (endTime - startTime) / 1_000_000.0;

        // Escrever números ordenados em um arquivo
        try (BufferedWriter writer = new BufferedWriter(new FileWriter("output.txt"))) {
            for (int num : numbers) {
                writer.write(num + "\n");
            }
        }

        // Obter o uso de memória
        long memoryUsageKb = (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024;

        // Exibir resultados
        System.out.printf("Tempo de execução: %.2f ms%n", executionTimeMs);
        System.out.printf("Uso de memória: %d KB%n", memoryUsageKb);
    }

    public static void main(String[] args) {
        String inputFile = "input.txt";

        try {
            measureMemoryAndTime(inputFile);
        } catch (IOException e) {
            System.err.println("Erro ao processar o arquivo: " + e.getMessage());
        }
    }
}

File: atividade2/merche.py
import time
import psutil  # type: ignore
import os

# Função para realizar o Merge Sort
def merge_sort(arr):
    if len(arr) < 2:
        return arr

    mid = len(arr) // 2

    # Dividir a lista
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    # Mesclar as metades ordenadas
    return merge(left, right)

def merge(left, right):
    sorted_arr = []
    i = j = 0

    # Mesclando os dois arrays
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            sorted_arr.append(left[i])
            i += 1
        else:
            sorted_arr.append(right[j])
            j += 1

    # Adicionando os elementos restantes
    sorted_arr.extend(left[i:])
    sorted_arr.extend(right[j:])
    return sorted_arr

# Função principal para medir tempo e uso de memória
def measure_memory_and_time(file_path):
    # Leitura do arquivo
    with open(file_path, "r") as file:
        numbers = [int(line.strip()) for line in file]

    # Medir o tempo inicial
    start_time = time.time()

    # Ordenar com Merge Sort
    sorted_numbers = merge_sort(numbers)

    # Medir o tempo final
    end_time = time.time()
    execution_time_ms = (end_time - start_time) * 1000  # Em milissegundos

    # Gravar a saída ordenada
    with open("output.txt", "w") as file:
        for num in sorted_numbers:
            file.write(f"{num}\n")

    # Medir memória usada
    process = psutil.Process(os.getpid())
    memory_usage_kb = process.memory_info().rss / 1024  # Em KB

    return execution_time_ms, memory_usage_kb

if __name__ == "__main__":
    input_file = "input.txt"
    execution_time, memory_usage = measure_memory_and_time(input_file)
    print(f"Tempo de execução: {execution_time:.2f} ms")
    print(f"Uso de memória: {memory_usage:.2f} KB")
